NullpointerException in MapValueCollection.writeData() in 3.1

We are in the process of upgrading our app from HZ 2.3 to 3.1. After upgrading a 2-node test cluster running our app to 3.1, I started up 1 node at a time. The first node came up fine and listed itself as the only cluster member. But when I started up the second node, it made a call to values() on an IMap, which eventually timed out after 2 minutes:

```
Caused by: com.hazelcast.core.OperationTimeoutException: No response for 120000 ms. Aborting invocation! InvocationFuture{invocation=InvocationImpl{ serviceName='hz:impl:mapService', op=com.hazelcast.map.operation.MapValuesOperation@57f897a7, partitionId=169, replicaIndex=0, tryCount=250, tryPauseMillis=500, invokeCount=1, callTim
eout=60000, target=Address[node1]:5701}, done=false}
    at com.hazelcast.spi.impl.InvocationImpl$InvocationFuture.waitForResponse(InvocationImpl.java:382) ~[hazelcast-3.1.jar:3.1]
    at com.hazelcast.spi.impl.InvocationImpl$InvocationFuture.get(InvocationImpl.java:294) ~[hazelcast-3.1.jar:3.1]
    at com.hazelcast.spi.impl.InvocationImpl$InvocationFuture.get(InvocationImpl.java:286) ~[hazelcast-3.1.jar:3.1]
    at com.hazelcast.spi.impl.OperationServiceImpl.invokeOnPartitions(OperationServiceImpl.java:552) ~[hazelcast-3.1.jar:3.1]
    at com.hazelcast.spi.impl.OperationServiceImpl.invokeOnAllPartitions(OperationServiceImpl.java:481) ~[hazelcast-3.1.jar:3.1]
    at com.hazelcast.map.proxy.MapProxySupport.valuesInternal(MapProxySupport.java:462) ~[hazelcast-3.1.jar:3.1]
    at com.hazelcast.map.proxy.MapProxyImpl.values(MapProxyImpl.java:426) ~[hazelcast-3.1.jar:3.1]
```

I then looked at the log for the first node, and saw the following:

```
INFO  2013-10-23 18:13:07,328 com.hazelcast.nio.SocketAcceptor  [node1]:5701 [v11] Accepting socket connection from /192.168.109.96:58266
INFO  2013-10-23 18:13:07,328 com.hazelcast.nio.TcpIpConnectionManager  [node1]:5701 [v11] 5701 accepted socket connection from /192.168.109.96:58266
INFO  2013-10-23 18:13:14,345 com.hazelcast.cluster.ClusterService  [node1]:5701 [v11]

Members [2] {
        Member [node1]:5701 this
        Member [node2]:5701
}

INFO  2013-10-23 18:13:14,362 com.hazelcast.partition.PartitionService  [node1]:5701 [v11] Re-partitioning cluster data... Migration queue size: 135
INFO  2013-10-23 18:13:16,215 com.hazelcast.partition.PartitionService  [node1]:5701 [v11] All migration tasks has been completed, queues are empty.
ERROR 2013-10-23 18:13:19,203 com.hazelcast.spi.impl.PartitionIteratingOperation  [node1]:5701 [v11] java.lang.NullPointerException
com.hazelcast.nio.serialization.HazelcastSerializationException: java.lang.NullPointerException
        at com.hazelcast.nio.serialization.SerializationServiceImpl.handleException(SerializationServiceImpl.java:290) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.nio.serialization.SerializationServiceImpl.writeObject(SerializationServiceImpl.java:246) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.nio.serialization.ByteArrayObjectDataOutput.writeObject(ByteArrayObjectDataOutput.java:257) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.spi.impl.PartitionIteratingOperation$PartitionResponse.writeData(PartitionIteratingOperation.java:129) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.nio.serialization.DataSerializer.write(DataSerializer.java:129) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.nio.serialization.DataSerializer.write(DataSerializer.java:36) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.nio.serialization.StreamSerializerAdapter.write(StreamSerializerAdapter.java:36) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.nio.serialization.SerializationServiceImpl.writeObject(SerializationServiceImpl.java:244) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.nio.serialization.ByteArrayObjectDataOutput.writeObject(ByteArrayObjectDataOutput.java:257) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.spi.impl.ResponseOperation.writeInternal(ResponseOperation.java:78) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.spi.Operation.writeData(Operation.java:273) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.nio.serialization.DataSerializer.write(DataSerializer.java:129) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.nio.serialization.DataSerializer.write(DataSerializer.java:36) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.nio.serialization.StreamSerializerAdapter.write(StreamSerializerAdapter.java:48) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.nio.serialization.SerializationServiceImpl.toData(SerializationServiceImpl.java:174) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.nio.serialization.SerializationServiceImpl.toData(SerializationServiceImpl.java:155) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.spi.impl.NodeEngineImpl.toData(NodeEngineImpl.java:151) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.spi.impl.OperationServiceImpl.send(OperationServiceImpl.java:579) [hazelcast-3.1.jar:3.1]
        at com.hazelcast.spi.impl.ResponseHandlerFactory$RemoteInvocationResponseHandler.sendResponse(ResponseHandlerFactory.java:112) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.spi.impl.OperationServiceImpl.doRunOperation(OperationServiceImpl.java:296) [hazelcast-3.1.jar:3.1]
        at com.hazelcast.spi.impl.OperationServiceImpl.access$800(OperationServiceImpl.java:52) [hazelcast-3.1.jar:3.1]
        at com.hazelcast.spi.impl.OperationServiceImpl$RemoteOperationProcessor.run(OperationServiceImpl.java:744) [hazelcast-3.1.jar:3.1]
        at com.hazelcast.util.executor.ManagedExecutorService$Worker.run(ManagedExecutorService.java:166) [hazelcast-3.1.jar:3.1]
        at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886) [na:1.6.0_35]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908) [na:1.6.0_35]
        at java.lang.Thread.run(Thread.java:662) [na:1.6.0_35]
        at com.hazelcast.util.executor.PoolExecutorThreadFactory$ManagedThread.run(PoolExecutorThreadFactory.java:59) [hazelcast-3.1.jar:3.1]
Caused by: java.lang.NullPointerException: null
        at com.hazelcast.map.MapValueCollection.writeData(MapValueCollection.java:49) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.nio.serialization.DataSerializer.write(DataSerializer.java:129) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.nio.serialization.DataSerializer.write(DataSerializer.java:36) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.nio.serialization.StreamSerializerAdapter.write(StreamSerializerAdapter.java:36) ~[hazelcast-3.1.jar:3.1]
        at com.hazelcast.nio.serialization.SerializationServiceImpl.writeObject(SerializationServiceImpl.java:244) ~[hazelcast-3.1.jar:3.1]
```

We never saw this issue with HZ 2.3.

I was able to work around the issue by making the following change to com.hazelcast.map.MapValueCollection's 1-param constructor:

``` java
    public MapValueCollection(Collection<Data> values) {
        this.values = new ArrayList<Data>(values.size());
        for (Data value : values) {
            if (value != null) {
                this.values.add(value);
            }
        }
    }
```

I don't know if this is the correct fix or if something earlier in the call path should be preventing the passed-in Collection from containing null values in the first place. 

I'd really appreciate it if you guys could sneak a fix for this into 3.1.1. Please let me know if I can provide any more details.

--- a/hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java
+++ b/hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java
@@ -34,7 +34,6 @@ public class GroupProperties {
     public static final String PROP_MEMCACHE_ENABLED = "hazelcast.memcache.enabled";
     public static final String PROP_REST_ENABLED = "hazelcast.rest.enabled";
     public static final String PROP_MAP_LOAD_CHUNK_SIZE = "hazelcast.map.load.chunk.size";
-    public static final String PROP_CACHED_NULL_TTL_SECONDS = "hazelcast.cached.null.ttl.seconds";
     public static final String PROP_MERGE_FIRST_RUN_DELAY_SECONDS = "hazelcast.merge.first.run.delay.seconds";
     public static final String PROP_MERGE_NEXT_RUN_DELAY_SECONDS = "hazelcast.merge.next.run.delay.seconds";
     public static final String PROP_OPERATION_CALL_TIMEOUT_MILLIS = "hazelcast.operation.call.timeout.millis";
@@ -111,8 +110,6 @@ public class GroupProperties {
     public final GroupProperty REST_ENABLED;
 
     public final GroupProperty MAP_LOAD_CHUNK_SIZE;
-
-    public final GroupProperty CACHED_NULL_TTL_SECONDS;
 
     public final GroupProperty MERGE_FIRST_RUN_DELAY_SECONDS;
 
@@ -226,7 +223,6 @@ public class GroupProperties {
         MEMCACHE_ENABLED = new GroupProperty(config, PROP_MEMCACHE_ENABLED, "true");
         REST_ENABLED = new GroupProperty(config, PROP_REST_ENABLED, "true");
         MAP_LOAD_CHUNK_SIZE = new GroupProperty(config, PROP_MAP_LOAD_CHUNK_SIZE, "1000");
-        CACHED_NULL_TTL_SECONDS = new GroupProperty(config, PROP_CACHED_NULL_TTL_SECONDS, "1");
         MERGE_FIRST_RUN_DELAY_SECONDS = new GroupProperty(config, PROP_MERGE_FIRST_RUN_DELAY_SECONDS, "300");
         MERGE_NEXT_RUN_DELAY_SECONDS = new GroupProperty(config, PROP_MERGE_NEXT_RUN_DELAY_SECONDS, "120");
         OPERATION_CALL_TIMEOUT_MILLIS = new GroupProperty(config, PROP_OPERATION_CALL_TIMEOUT_MILLIS, "60000");

--- a/hazelcast/src/main/java/com/hazelcast/map/DefaultRecordStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/DefaultRecordStore.java
@@ -469,15 +469,6 @@ public class DefaultRecordStore implements RecordStore {
                     saveIndex(record);
                     updateSizeEstimator(calculateRecordSize(record));
                 }
-                // below is an optimization. if the record does not exist the next get will return null without looking at mapStore.
-                if (value == null) {
-                    int ttlForNull = mapService.getNodeEngine().getGroupProperties().CACHED_NULL_TTL_SECONDS.getInteger();
-                    if (ttlForNull > 0) {
-                        record = mapService.createRecord(name, dataKey, null, ttlForNull * 1000);
-                        records.put(dataKey, record);
-                        updateSizeEstimator(calculateRecordSize(record));
-                    }
-                }
             }
 
         } else {
@@ -525,15 +516,6 @@ public class DefaultRecordStore implements RecordStore {
                 saveIndex(record);
                 updateSizeEstimator(calculateRecordSize(record));
             }
-            // below is an optimization. if the record does not exist the next get will return null without looking at mapStore.
-            else {
-                int ttlForNull = mapService.getNodeEngine().getGroupProperties().CACHED_NULL_TTL_SECONDS.getInteger();
-                if (ttlForNull > 0) {
-                    Record record = mapService.createRecord(name, dataKey, null, ttlForNull * 1000);
-                    records.put(dataKey, record);
-                    updateSizeEstimator(calculateRecordSize(record));
-                }
-            }
             value = mapService.interceptGet(name, value);
             if (value != null) {
                 mapEntrySet.add(new AbstractMap.SimpleImmutableEntry(dataKey, mapService.toData(value)));
@@ -556,9 +538,7 @@ public class DefaultRecordStore implements RecordStore {
             }
         }
 
-        // because of a get optimization (see above), there may be a record with a null value,
-        // which means map-store returned null while loading the key.
-        boolean contains = record != null && record.getValue() != null;
+        boolean contains = record != null;
         if (contains) {
             accessRecord(record);
         }
@@ -570,7 +550,7 @@ public class DefaultRecordStore implements RecordStore {
         Data dataKey = entry.getKey();
         Object value = entry.getValue();
         Record record = records.get(dataKey);
-        if (record == null || record.getValue() == null) {
+        if (record == null) {
             value = mapService.interceptPut(name, null, value);
             record = mapService.createRecord(name, dataKey, value, -1);
             mapStoreWrite(record, dataKey, value);
@@ -602,7 +582,7 @@ public class DefaultRecordStore implements RecordStore {
         checkIfLoaded();
         Record record = records.get(dataKey);
         Object oldValue = null;
-        if (record == null || record.getValue() == null) {
+        if (record == null) {
             if (mapContainer.getStore() != null) {
                 oldValue = mapContainer.getStore().load(mapService.toObject(dataKey));
             }
@@ -638,7 +618,7 @@ public class DefaultRecordStore implements RecordStore {
         checkIfLoaded();
         Record record = records.get(dataKey);
         boolean newRecord = false;
-        if (record == null || record.getValue() == null) {
+        if (record == null) {
             value = mapService.interceptPut(name, null, value);
             record = mapService.createRecord(name, dataKey, value, ttl);
             mapStoreWrite(record, dataKey, value);
@@ -664,7 +644,7 @@ public class DefaultRecordStore implements RecordStore {
         checkIfLoaded();
         Record record = records.get(dataKey);
         Object newValue = null;
-        if (record == null || record.getValue() == null) {
+        if (record == null) {
             newValue = mergingEntry.getValue();
             record = mapService.createRecord(name, dataKey, newValue, -1);
             mapStoreWrite(record, dataKey, newValue);
@@ -718,7 +698,7 @@ public class DefaultRecordStore implements RecordStore {
     public boolean replace(Data dataKey, Object testValue, Object newValue) {
         checkIfLoaded();
         Record record = records.get(dataKey);
-        if (record == null || record.getValue() == null)
+        if (record == null)
             return false;
         if (mapService.compare(name, record.getValue(), testValue)) {
             newValue = mapService.interceptPut(name, record.getValue(), newValue);
@@ -737,7 +717,7 @@ public class DefaultRecordStore implements RecordStore {
     public void putTransient(Data dataKey, Object value, long ttl) {
         checkIfLoaded();
         Record record = records.get(dataKey);
-        if (record == null || record.getValue() == null) {
+        if (record == null) {
             value = mapService.interceptPut(name, null, value);
             record = mapService.createRecord(name, dataKey, value, ttl);
             records.put(dataKey, record);
@@ -754,7 +734,7 @@ public class DefaultRecordStore implements RecordStore {
 
     public void putFromLoad(Data dataKey, Object value, long ttl) {
         Record record = records.get(dataKey);
-        if (record == null || record.getValue() == null) {
+        if (record == null) {
             value = mapService.interceptPut(name, null, value);
             record = mapService.createRecord(name, dataKey, value, ttl);
             records.put(dataKey, record);
@@ -772,7 +752,7 @@ public class DefaultRecordStore implements RecordStore {
     public boolean tryPut(Data dataKey, Object value, long ttl) {
         checkIfLoaded();
         Record record = records.get(dataKey);
-        if (record == null || record.getValue() == null) {
+        if (record == null) {
             value = mapService.interceptPut(name, null, value);
             record = mapService.createRecord(name, dataKey, value, ttl);
             mapStoreWrite(record, dataKey, value);
@@ -795,7 +775,7 @@ public class DefaultRecordStore implements RecordStore {
         checkIfLoaded();
         Record record = records.get(dataKey);
         Object oldValue = null;
-        if (record == null || record.getValue() == null) {
+        if (record == null) {
             if (mapContainer.getStore() != null) {
                 oldValue = mapContainer.getStore().load(mapService.toObject(dataKey));
                 if (oldValue != null) {

--- a/hazelcast/src/main/java/com/hazelcast/map/operation/MapReplicationOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/operation/MapReplicationOperation.java
@@ -73,7 +73,6 @@ public class MapReplicationOperation extends AbstractOperation {
                 Data key = recordEntry.getKey();
                 Record record = recordEntry.getValue();
                 if (record.getValue() == null) {
-                    // see optimization at DefaultRecordStore.get(Data dataKey)
                     continue;
                 }
                 RecordReplicationInfo recordReplicationInfo;

--- a/hazelcast/src/test/java/com/hazelcast/map/mapstore/MapStoreTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/mapstore/MapStoreTest.java
@@ -764,7 +764,6 @@ public class MapStoreTest extends HazelcastTestSupport {
         final ConcurrentMap<String, String> store = new ConcurrentHashMap<String, String>();
         final MapStore<String, String> myMapStore = new SimpleMapStore<String, String>(store);
         Config config = new Config();
-        config.setProperty(GroupProperties.PROP_CACHED_NULL_TTL_SECONDS, "0");
         config
                 .getMapConfig("testIssue806CustomTTLForNull")
                 .setMapStoreConfig(new MapStoreConfig()
@@ -810,6 +809,24 @@ public class MapStoreTest extends HazelcastTestSupport {
         map.put("key", "value");
         Thread.sleep(2000);
         assertEquals("value", map.get("key"));
+    }
+
+    @Test
+    public void testIssue1019() {
+        Config config = new Config();
+        MapStoreConfig mapStoreConfig = new MapStoreConfig();
+        mapStoreConfig.setImplementation(new MapStoreAdapter<String,String>());
+        config.getMapConfig("map").setMapStoreConfig(mapStoreConfig);
+        HazelcastInstance instance = Hazelcast.newHazelcastInstance(config);
+        HazelcastInstance instance2 = Hazelcast.newHazelcastInstance(config);
+        final IMap map = instance.getMap("map");
+        for (int i = 0; i < 1000; i++) {
+            map.put(i,i);
+        }
+        for (int i = 10000; i < 10100; i++) {
+            map.get(i);
+        }
+        assertEquals(1000, map.values().size());
     }
 
     public static Config newConfig(Object storeImpl, int writeDelaySeconds) {

